WIP: Improve accuracy of ^ for ill-conditioned matrices

mfasi opened the pull request:
Add the algorithm for `A^t` from Higham and Lin, 2013. See JuliaLang/julia#5840.

I split the old version of `logm` into several functions to reuse some of the code.

mfasi committed:
Make the algorithm for real powers of a matrix robust

StefanKarpinski commented:
"Improve" is a little vague. Does this change behavior or just performance? Some explanation and examples of the difference would be nice.

mfasi commented:
@StefanKarpinski You're right, an explanation might help. The implementation that we have now in `base` is not reliable for general matrices because it fails when the eigenvectors of the matrix are ill-conditioned. A classical example is the comparison of the (relative) residuals for
```javascript
julia> e= 1e-16; A = [1+e 1; 0 1];
julia> cond(A)
2.618033988749896
julia> F = eigfact(A); cond(F.vectors)
9.007199254740992e15
```
On `master` I get
```javascript
julia> norm((A^(1/2))^2 - A) / norm(A)
0.6180339887498948
```
and on `powm`
```javascript
julia> norm((A^(1/2))^2 - A) / norm(A)
6.861555643110582e-17
```

To answer your question, I'd say that the focus is on the behaviour here.

stevengj commented on base/linalg/dense.jl:
```diff
@@ -169,17 +169,63 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
+        warn("Matrix with nonpositive real eigenvalues, a nonprincipal matrix power will be returned.")
+    end
+
+    if isreal(A) && ~np_real_eigs
```
Are we worried about type-stability of this function?  Or is it slow enough that we don't care?

StefanKarpinski replied:
If we can be type stable without sacrificing too much convenience, it's better to, even if things aren't performance critical.

andreasnoack replied:
So far we haven't been type stable for matrix functions like this, e.g. `eig` and `sqrtm` return `Float64` for `Float64` input if possible. If we want type stability, we could

1. Mimic scalar function and return `DomainError`
2. Always return `Complex`

I think option 1 is a no go because the consequence would be to solve real eigenvalue problems in complex arithmetic which is much more expensive. The main problem with option two is that some people might be annoyed by the complex return matrix when all entries are real valued, but that might be a reasonable price for type stability.

stevengj commented on base/linalg/dense.jl:
```diff
@@ -169,17 +169,63 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
+        return powm(A,p1) * A^floor(p)
+    else
+        return powm(A,p - ceil(p)) * A^ceil(A)
```
`A^ceil(A)` doesn't look right.

(Be sure that the test suite covers all of these branches!)

mfasi replied:
You spot another bug, thank you!

stevengj commented on base/linalg/dense.jl:
```diff
@@ -169,17 +169,63 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
+
+    # General matrices
+    p1 = p - floor(p)
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
```
Note that `cond(A)` is relatively expensive; nearly half as expensive as `schurfact` on my machine.  Can't we push this check into `powm`, and get (or estimate) the condition number more quickly once we have the Schur factorization?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,71 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+    mpm = 2 * m
+    c = 0.5 * (p - m) / (mpm - 1)
+    S = c * A
+    for j = m-1:-1:1
```
It seems like this loop makes a lot of copies of matrices.  Can't the scaling of `S` be done in-place?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,71 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+
+    blockpower!(A0,S,p/(2^s))
+    for i = 1:s
+        S = UpperTriangular(S * S)
```
Again a lot of matrix copies here; maybe use `A_mul_B!` with pre-allocated output, and then update `S` in-place from the result?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    # Compute the Padé approximation
+    Y = zeros(T, n, n)
+    for k = 1:m
+        Y = Y + w[k] * (A / (x[k] * A + I))
```
Again a lot of matrix allocations here; would be better to replace with in-place updates of `Y` etc.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    for i = 1:n
+        a = complex(A0[i,i])
+        if s == 0
+            r = a - 1
```
this value of `r` is never used?

mfasi replied:
You found a bug, thank you!

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+
+# Auxiliary functions for logm and matrix power
+
+function sqrt_diag!(A0::UpperTriangular, A::UpperTriangular, s)
```
Some comment on what this function actually computes would be nice...

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+            a = sqrt(a)
+            r = r * (1 + a)
+        end
+        A[i,i] = z0 / r
```
If `A` is a real matrix, won't this potentially throw an `InexactError` since `z0` is complex?  And if it doesn't, why do you need `a = complex(A0[i,i])` rather than `a = A0[i,i]`?

mfasi replied:
`A` can't be real, it's a (complex) copy of `A0`. I think it handles the complex case as well.

stevengj replied:
In that case, this constraint should be specified in the type signature.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+            r = a - 1
+        end
+        s0 = s
+        if angle(a) >= pi / 2
```
This check seems odd to me.  Do you just want to check for `real(a) < 0`?

mfasi replied:
Not sure, but I think we want to see when we jump on the other side of the branch cut, which the negative real line.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
-    n = size(A0, 1)
-    A = copy(A0)
+    n = chksquare(A0)
+    A = copy(complex(A0))
```
Why the conversion to complex?

mfasi replied:
I think it's needed because the p-th root of a negative real number can be complex.

stevengj replied:
Yes, but you don't actually compute any roots until the call to `blockpower` at the end.  It can stay real (if `A0` is real) until then, no?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1095,77 +1244,19 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    return (isreal(x)? zero(T) : T(ceil((imag(x) - pi) / (2 * pi))))
 end
-logm(A::LowerTriangular) = logm(A.').'
+@vectorize_1arg Number unw
```
Why is there a vectorized version of this function?  We don't seem to use it, and the function is not exported.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1095,77 +1244,19 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
-    return UpperTriangular(Y)
-
+function unw{T<:Number}(x::T)
+    return (isreal(x)? zero(T) : T(ceil((imag(x) - pi) / (2 * pi))))
```
Maybe:
```
unw(x::Real) = 0
unw(x::Number) = ceil((imag(x) - pi) / (2 * pi))
```
I don't see the point in casting to `T`.

mfasi commented:
@stevengj Thank you for checking the patch thoroughly. I addressed most of the points you raised, still to do:
* add tests
* add documentation
* push `cond(A)` after the Schur decomposition
I'll make this PR a WIP and will probably add one or two other commits (tests and docs). Not sure I found the least expensive way to do the in place operations, so feel free to check the loops again (I'll mark them with a line comment for convenience).

mfasi commented:
I managed to push the use of `cond(A)` after the Schur factorization. I added inline documentation and tests as well.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
     end
 end
 
+function ^(A::UpperTriangular, p::Integer)
```
I think it would probably be clearer, and consistent with standard practice in linear algebra, to name your upper-triangular matrices `U` instead of `A`.

mfasi replied:
Agreed, but I'd rather prefer not to do it with this PR, because it would break the consistency with the notation of `sqrtm`. I'll put it in my TODO list.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+    end
+    # Compute the Padé approximant
+    c = 0.5 * (p - m) / (2 * m - 1)
+    S = c * A
```
`S` is not type stable: `c*A` returns an upper-triangular matrix, but it looks like `S \ A` returns a full matrix.  The latter should probably be fixed, since the product and inverse of upper-triangular matrices are upper-triangular.

mfasi replied:
The same problem I experienced with the matrix product: the product of two `UpperTriangular` matrices is not `UpperTriangular`. I guess there must be a reason for this (maybe related to the underlying BLAS kernels), but I'd be happy to work on the additional methods. Would it be worth it to open an issue and discuss that?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+        for i = 1:n
+            S[i,i] = S[i,i] + 1
+        end
+        S = S \ A
```
To avoid repeated matrix allocations, maybe define a matrix `Stmp = similar(S)` and call `triu!(A)` outside the loop, then replace `S = S \ A` by:
```jl
copy!(Stmp.data, S.data)
copy!(S.data, A.data)
A_ldiv_B!(Stmp, S.data) # S = S \ A
```

stevengj replied:
You could also eliminate the `scale!(c, S)` line by replacing `copy!(S.data, A.data)` above by `scale!(S.data, A.data, c)`

mfasi replied:
Not sure we need `triu!(A)`, as `A` should be `UpperTriangular` already. Also, did you mean to say that I can substitute `copy!(Stmp.data, S.data)` with `scale!(Stmp.data, S.data, c)`? I don't understand why it should work otherwise.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+    S = UpperTriangular(S)
+
+    blockpower!(A0,S,p/(2^s))
+    D = Array(typeof(S[1,1]),n,n)
```
Why not `Array(T,n,n)`?

stevengj replied:
Or `D = Stmp.data`, given the code above.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+    D = Array(typeof(S[1,1]),n,n)
+    for m = 1:s
+        A_mul_B!(D,S,S)
+        for j = 1:n
```
Might be faster to just do `copy!(S.data, D)`; this copies twice as much data, but can take advantage of contiguous copies etc.

stevengj replied:
Nope, I just checked, it is slower to do `copy!`.

We should really pull this code out into a `copy!` method for triangular matrices.  See also the `copy!` usages in my code above, which would be cleaner with a `copy!` method.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+        A_mul_B!(D,S,S)
+        for j = 1:n
+            for i = 1:j
+                S[i,j] = D[i,j]
```
`@inbounds`

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -976,18 +976,110 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+        end
+        blockpower!(A0,S,p/(2^(s-m)))
+    end
+    if isreal(S)
```
maybe `T <: Real && `typeof(p) <: Integer`?  Then it would be type-stable.

mfasi replied:
This function called with `p` in (-1,0) or (0,1). I simply removed everything and put 
```
 return S
```

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+                D[i,j] = A[i,j]
+            end
+        end
+        scale!(x[k], D)
```
See above: a `scale!(c, U1, U2)` for upper-triangular matrices would come in handy here.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+
+    # Compute the Padé approximation
+    Y = zeros(T, n, n)
+    D = zeros(T, n, n)
```
Why aren't `Y` and `D` upper-triangular?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+        for i = 1:n
+            D[i,i] = D[i,i] + 1
+        end
+        D = (A / D)
```
See above: we want to (a) avoid repeated allocations of new matrices and (b) take advantage of the fact(?) that `D` is upper triangular.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+        scale!(w[k],D)
+        for j = 1:n
+            for i = 1:j
+                Y[i,j] = Y[i,j] + D[i,j]
```
`+=`?

mfasi replied:
Thank you for raising this point. Are there guidelines on the use of abbreviations? I've never used them so far, are there preferred in general?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+        elseif 2 * abs(Ak) < abs(Akp1) || 2 * abs(Akp1) < abs(Ak)
+            Y[k,k+1] = A0[k,k+1] * (logAkp1 - logAk) / (Akp1 - Ak)
+        else
+            w = atanh((Akp1 - Ak)/(Akp1 + Ak) + im * pi * unw(logAkp1-logAk))
```
do we have test coverage for all of these branches?

mfasi replied:
Yes, the matrix logarithm was already there. Coveralls guarantees that all these lines are covered.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    end
+end
+
+# Compute the repeated square root of A to match the conditions on theta
```
This comment is a little inscrutable.  What conditions on theta?

Note also that you can name the variable `θ` (`\theta<TAB>`) rather than `theta`.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+
+# Auxiliary functions for logm and matrix power
+
+# Compute accurate diagonal of A = A0^s - I
```
is there a reference for the algorithm?

mfasi replied:
Yes, there is. I added it.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -995,15 +1087,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+            A[i,i] = a - 1
+        else
+            s0 = s
+            if angle(a) >= pi / 2
```
`imag(a) >= 0 & real(a) <= 0` is equivalent but faster

mfasi commented:
@stevengj I think I addressed all your points, it took some time because I had to implement `scale!` and `copy!` for triangular matrices.

stevengj commented on base/linalg/symmetric.jl:
```diff
@@ -174,7 +174,54 @@ function svdvals!{T<:Real,S}(A::Union{Hermitian{T,S}, Symmetric{T,S}, Hermitian{
+    end
+end
+function ^(A::Symmetric, p::Real)
+    F = eigfact(full(A))
```
This seems problematic for complex-symmetric matrices, which may be nearly defective.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1087,15 +1174,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    sqrt_diag!(A0, A, s)
+
+    # Compute the Gauss-Legendre quadrature formula
+    R = zeros(Float64, m, m)
```
Use `Base.QuadGK.gauss`?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1087,15 +1174,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+        R[i,i+1] = i / sqrt((2 * i)^2 - 1)
+        R[i+1,i] = R[i,i+1]
+    end
+    x,V = eig(R)
```
(If you were to do this here, you would probably want to use a `SymTridiagonal` matrix `R`, no?)

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1068,18 +1068,105 @@ for (f, g) in ((:/, :A_rdiv_B!), (:A_rdiv_Bc, :A_rdiv_Bc!), (:A_rdiv_Bt, :A_rdiv
+        ArgumentError("p must be a real number in (-1,1), got $p")
+    end
+
+    theta = [1.53e-5, 2.25e-3, 1.92e-2, 6.08e-2, 1.25e-1, 2.03e-1, 2.84e-1]
```
What are these numbers?  Can they be computed easily in a precision-independent fashion so that this function is not restricted to `Float64`?

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1087,15 +1174,116 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
          1.276404810806775e-001,
          2.060962623452836e-001,
          2.879093714241194e-001]
+
```
Same as above: what is `theta`, and can we make this routine precision-independent?

mfasi replied:
Not sure it is practical to compute them each time (actually I'm pretty sure it isn't), maybe we could tabulate them for different precision framework.

stevengj replied:
How are they computed?   If the computation is expensive, we could always cache them like we do the Gauss-Kronrod weights in quadgk.jl

mfasi replied:
The `i`-th entry is the maximal value of `||X||` such that `||X||<1` and a complicate expression involving `X` and `i` is smaller than our current `eps`. I have no idea how long it would take to compute them for a given precision, but it might be matter of minutes, that's why I suggested to store them hard coded.
I'll try to find out how these values can be computed and fix the other issues you pointed out in the meanwhile.

mfasi commented:
@stevengj A [gist](https://gist.github.com/mfasi/1d01bcc6d39a5ea14309) of the functions to compute `theta` for `logm` and `powm`. I probably used the `BigFloat` constructor more than I had to, but the code seems OK. The two functions run in more than 30 seconds, less than a minute on my machine. What do you suggest I do to integrate them?

mfasi commented:
It looks like I was wrong, to have multiprecision algorithms you don't need the `theta`s only, but you have to modify the algorithm as well. As those change are far from being trivial and I don't think I can figure them out, I'd say that the patch is ready. Any comments?

stevengj commented:
Can you squash the commits?

stevengj commented:
Can the existing algorithm and theta values be used in single precision, at least?

mfasi commented:
@stevengj I think it wouldn't return a wrong answer for single precision, but could potentially do a lot of non necessary work.

stevengj commented:
@mfasi, as long as it does less unnecessary work than converting the whole matrix to double precision, with little or no expansion in code size, it's still a win.  We can always optimize it letter.

alyst commented on base/linalg/dense.jl:
```diff
@@ -169,17 +169,50 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
+    # Check whether the matrix has nonpositive real eigs
+    np_real_eigs = false
+    for i = 1:n
+        if imag(d[i]) < eps() && real(d[i]) <= 0
```
Maybe `abs(imag(d[i])) < eps() && real(d[i]) < -eps()`?

alyst commented on base/linalg/diagonal.jl:
```diff
@@ -161,9 +161,14 @@ end
-logm(D::Diagonal) = Diagonal(log(D.diag))
-sqrtm(D::Diagonal) = Diagonal(sqrt(D.diag))
+# Matrix functions
+^(D::Diagonal, p::Integer) = Diagonal((D.diag).^p)
```
`^(D::Diagonal, p::Integer)` method is not necessary since its code is identical to `p::Real` case.

alyst commented on base/linalg/dense.jl:
```diff
@@ -169,17 +169,50 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
+
+    # Check whether the matrix has nonpositive real eigs
+    np_real_eigs = false
+    for i = 1:n
```
IIRC, `for i = 1:n` syntax is deprecated in favour of `for i in 1:n`

tkelman replied:
No it is not.

alyst commented on base/linalg/symmetric.jl:
```diff
@@ -198,8 +198,55 @@ function svdvals!{T<:Real,S}(A::Union{Hermitian{T,S}, Symmetric{T,S}, Hermitian{
+        if T <: Real
+            return Hermitian(retmat)
+        else
+            for i = 1:n
```
`for i in 1:n`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    # General UpperTriangular matrices
+    p1 = p - floor(p)
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
+        return powm(A,p1) * A^floor(p)
```
Note that `floor(p)` has the same type as `p`. Here, `floor(Int, p)` should be better as it would trigger the integer version of `M^p` directly.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
+        return powm(A,p1) * A^floor(p)
+    else
+        return powm(A,p - ceil(p)) * A^ceil(p)
```
`A^ceil(Integer, p)`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    for j = m-1:-1:1
+        j4 = 4 * j
+        c = (-p - j) / (j4 + 2)
+        for i = 1:n
```
`for i in 1:n`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+        A_ldiv_B!(Stmp,S.data)
+
+        c = (p - j) / (j4 - 2)
+        for i = 1:n
```
`for i in 1:n`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+        scale!(S,A,c)
+        A_ldiv_B!(Stmp,S.data)
+    end
+    for i = 1:n
```
`for i in 1:n`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    copy!(Stmp,S)
+    scale!(S,A,-p)
+    A_ldiv_B!(Stmp,S.data)
+    for i = 1:n
```
`for i in 1:n`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    end
+
+    blockpower!(A0,S,p/(2^s))
+    for m = 1:s
```
`for m in 1:s`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+
+    # Compute the Gauss-Legendre quadrature formula
+    x, w = Base.QuadGK.gauss(Float64, m)
+    for i = 1:m
```
`for i in 1:m`

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    x, w = Base.QuadGK.gauss(Float64, m)
+    for i = 1:m
+        x[i] += 1
+        x[i] *= 0.5
```
`@inbounds x[i] = 0.5*(x[i]+1)`?

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1474,91 +1660,45 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+end
+
+# Compute accurate diagonal and superdiagonal of A = A0^p
+function blockpower!(A0::UpperTriangular, A::UpperTriangular, p)
```
By convention, `xyz!(y, a, b)` methods are expected to modify the first argument. See e.g. `A_mul_B!(y, A, B)`. It's better to swap the order of `A` and `A0`.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
-    d = diag(A)
-    dm1 = Array(T, n)
+    d = complex(diag(A))
+    dm1 = Array(typeof(d[1,1]), n)
```
`Vector{eltype(d)}(n)`?

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1440,6 +1623,7 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
             end
             if j <= 6
                 m = j
+                foundm = true
```
Since `foundm` is not really used in this loop and afterwards, I'm tempting to suggest to eliminate it completely, e.g.:
```julia
if alpha2 <= theta[2]
    m = alpha2 <= theta[1] ? 1 : 2
else
    while true
       <...>
    end
end
```
We would save a few lines + improve (IMHO) the clarity at the expense of indentation.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
     tmax = size(theta, 1)
-    n = size(A0, 1)
+    n = checksquare(A0)
     A = copy(A0)
```
`complex(copy(A0))` since its converted into complex later anyway?

alyst replied:
Would it make preemptively define `A` as complex: `A = complex(copy(A0))`? As later in the loop we have `A = sqrtm(A)`, it should simplify `A` type inference and potentially make the things faster.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+
+# Repeatedly compute the square roots of A so that in the end its
+# eigenvalues are close enough to the positive real line
+function invsquaring{T}(A0::UpperTriangular{T}, theta)
```
It looks like `T` is not really necessary.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1474,91 +1660,45 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+# Compute accurate diagonal and superdiagonal of A = A0^p
+function blockpower!(A0::UpperTriangular, A::UpperTriangular, p)
+    n = checksquare(A0)
     for k = 1:n-1
```
`@inbounds for k in 1:(n-1)`?

alyst commented:
@mfasi I've come across the limitations of the current implementation in #16930. It would be really nice to have your great work finally merged. Could you please rebase it to the current master?

MatthiasJReisinger committed:
Add command line option --polly={yes|no}

Passing `--polly=no` will cause `@polly` declarations to be ignored.
This facilitates debugging or evaluating performance differences between
using/not using Polly without having to manually remove `@polly`
declarations from functions.

alyst commented on base/linalg/triangular.jl:
```diff
@@ -1369,34 +1369,217 @@ end
+    scale!(2^s,Y.data)
+
+    # Compute accurate diagonal and superdiagonal of log(A)
+    for k = 1:n-1
```
`@inbounds for k in 1:(n-1)`

tkelman commented on base/linalg/dense.jl:
```diff
@@ -150,17 +150,50 @@ kron(a::AbstractVector, b::AbstractVector)=vec(kron(reshape(a,length(a),1),resha
-    (X * Diagonal(v.^p)) * Xinv
+
+    # If possible, use diagonalization
+    if issymmetric(A) && T <: Real
```
switch the order here so complex matrices don't get double-checked once for symmetry then again for hermitian-ness?

tkelman commented on base/linalg/dense.jl:
```diff
@@ -264,7 +297,7 @@ function rcswap!{T<:Number}(i::Integer, j::Integer, X::StridedMatrix{T})
 
 """
-    logm(A::StridedMatrix)
+    logm(A{T}::StridedMatrix{T})
```
you'll need to make the corresponding change in the rst docs

Sacha0 committed:
Rewrite vectorized unary functions over SparseMatrixCSCs, leveraging higher order functions and multiple dispatch to displace eval. Fixes some apparent type instabilities.

Sacha0 committed:
Transition vectorized unary functions over `SparseMatrixCSC`s to compact broadcast syntax, accordingly revise and expand the associated tests, and add deprecations for the vectorized syntax.

JeffBezanson committed:
sweep unused types from caches in sysimg write

Sacha0 committed:
Deprecate vectorized functions in base/special/trig.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/log.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/gamma.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/erf.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/special/bessel.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/math.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/floatfuncs.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/fastmath.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/complex.jl in favor of compact broadcast syntax.

JeffBezanson committed:
avoid making copies of `Tuple{Vararg{Any}}`

Sacha0 committed:
Deprecate vectorized functions in base/dates/accessors.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/adjusters.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/conversions.jl in favor of compact broadcast syntax.

Sacha0 committed:
Deprecate vectorized functions in base/dates/query.jl in favor of compact broadcast syntax.

Sacha0 committed:
Clean up @vectorize_(1|2)arg deprecations.

tkelman committed:
Delete some files from libssh2 and curl that keep changing on travis

causing the cache to get re-uploaded

yuyichao committed:
Add more `try`-`catch` and sigatomic for top-level code/new tasks

So that we don't need to run `jl_exit` in strange (signal handler) context
due to missing exception handler.

yuyichao committed:
Implement `jl_call_in_ctx` on unix.

Use it to make sure that `jl_rethrow` and `jl_exit` are running on the right
thread and right stack when an exception/exit is caused by a signal.

Fix #17706

yuyichao committed:
Optimize exception frames allocation

* Make llvm-gcroot a ModulePass
* Use llvm-gcroot to calculate the minimum number of exception frames necessary.

maleadt committed:
Fix `@threadcall` argument passing.

Fixes #17819. Also document not being allowed to call back into Julia.

tkelman committed:
Make rpath settings a bit more consistent

probably addresses #18106

vtjnash committed:
add mem fence-post verification code

defining MEMFENCE will tell the gc to validate all object tags
(before they get freed) to attempt to detect out-of-bounds writes
that corrupted a object tag or the freelist

p-zubieta committed:
Improve inferability of promote_op

timholy committed:
Support non-1 indices and fix type problems in DFT (fixes #17896)

timholy committed:
Export circcopy! and add it to the RST manual

tkelman committed:
Revert "Temporary ugly hack to download old versions of winrpm gcc dll's" (#17906)

This reverts commit 97f18418b257bdcf625a5384687389b048e9c2ab.
ref #15521

wildart committed:
fix incorrect bignum rng buffer size [fix #17772] (#17874)

tkelman committed:
Patch is picky about newlines at end of file

Keno committed:
Set an RPATH on all libraries requiring fortran in binary-dist (#17901)

andreasnoack committed:
Remove redundant uplo argument in chol family. When using Hermitian (#17909)

and Symmetric, it is redundant to also have an uplo argument and
occasionally, it also gave the wrong result. This can therefore be
considered a bugfix.

andreasnoack committed:
Define conj and conj! for for Symmetric and Hermitian. (#17827)

Fixes #17780

ranjanan committed:
Add cross-release compatibility note to CONTRIBUTING.md (#17911)

* Add cross-release compatibility note to CONTRIBUTING.md

* tag -> specify

tkelman committed:
reword second tag -> find in CONTRIBUTING.md

ref #17911 [ci skip]

KristofferC committed:
improve heuristic for when to refresh line in the REPL (#17868)

* improve heuristic for when to refresh line

* fix trailing whitespace...

* use indent when not on first line

* remove unused argument and change local var to not share name with function

MichaelHatherly committed:
Fix trailing whitespace in code literal

rst apparently requires that the closing double backticks for code literals
appear at the end of a "word", rather than preceded by a space as markdown
allows. This means that ``` ``julia> `` ``` doesn't parse correctly. The
fix is to just remove the trailing space and explicitly mention the need
for a space after the `>`.

vtjnash committed:
preserve functionality of special ^ inliner in linear IR

JeffBezanson committed:
fix missing line break in precompile error message

JeffBezanson committed:
remove some unnecessary anonymous functions

Transform `x->f(x)` to `f`.

vtjnash committed:
add model of special case ^ inliner to inference edges

required for the inliner to work, since all of the logic for computing edges is supposed to be handle by the inference step

fix #17759

JeffBezanson committed:
fix BigInt `ndigits` test

See #16766. `ndigits(big(0),1)` doesn't raise an error, but we were
sometimes (randomly, rarely) testing that it does. This change makes
the test reliable. The issue of how this case should actually
behave is still open.

vtjnash committed:
improve correctness of fieldtype_tfunc

Vararg is only exact if in covariant position

also make getfield_tfunc monotonic for the case where the type has one
field, to avoid the same bug

fix #16530

timholy committed:
Move OffsetArray definition to TestHelpers

Commit:
fix output of getaddrinfo in networking and streams (#17931)

fix output of getaddrinfo in networking and streams documentation for v0.5.

tkelman committed:
Fix doctest line numbers after #17882

(cherry picked from commit 84ad7c720cb0aaaecb226388eb9c00a5a2ba2041)
[ci skip]

ranjanan committed:
Add test for PR #17803

The PR fixes jl_static_show for bitstypes

vtjnash committed:
add support for split debuginfo files

There seem to be a few variations on the tools and arguments
available for creating split debuginfo files.
I used the following:

objcopy --only-keep-debug libjulia.so libjulia.so.debug
strip -g libjulia.so
objcopy --add-gnu-debuglink=libjulia.so.debug libjulia.so

yuyichao committed:
Fix typo in tbaa_decorate_gcframe

tkelman committed:
Add cygpath_w call so `make -C test/perf` works in cygwin-cross

timholy committed:
Fix StackOverflowError in zip_iteratorsize

JeffBezanson committed:
slightly improve error message for unsupported kw arguments

also fix some code formatting

yuyichao committed:
Fix MSVC compilation

Also update out-of-date comment about codegen steps

ranjanan committed:
Change test to Int128(-1)

MichaelHatherly committed:
Remove two cross-references

  * Replace a malformed link in `write` docstring with a literal `write` rather than
    a real cross-reference, since that would just cross-reference itself. Also move
    `write` docstring inline.

  * Remove cross-reference for `<<` from inside a literal since text inside a literal
    can't contain a link. A cross-reference to `<<` already appears at the end of the
    docstring.

tkelman committed:
fix deprecated syntax "global rng_state::Float64"

timholy committed:
Expand docs on enumerate's relationship with indexing

timholy committed:
Generalize getindex for AbstractUnitRanges

It turns out we were returning a Vector{Int} for operations that should return a Range

andreasnoack committed:
Revert "Remove redundant uplo argument in chol family. When using Hermitian (#17909)"

This reverts commit d992f3dd10602ef52d8c297ebb7c7396a579c963.

andreasnoack committed:
Remove redundant uplo argument in chol family. When using Hermitian
and Symmetric, it is redundant to also have an uplo argument and
occasionally, it also gave the wrong result. This can therefore be
considered a bugfix.

wildart committed:
disable openssl in curl build when build agains mbedtls [fix #17910] (#17955)

xorJane committed:
Move documentation inline for error types declared in base/base.jl. (#17959)

Keno committed:
Fix typo in libgit2 credentials callback (#17975)

StefanKarpinski committed:
more manual eta reduction (companion to 9cead4f7874a2aa9b1a8d6adc5) (#17982)

kshyatt committed:
Fix atahn to atanh (#17977)

* Fix atahn to atanh

* More tests

ranjanan committed:
Write up test for #15592

parse("[1] [2]") should throw a ParseError

Commit:
Remove nonstandard header sys/sysctl.h when unneeded

This header prevents me from building Julia on musl libc.

```
/home/alba/julia/src/sys.c:16:24: fatal error: sys/sysctl.h: No such file or directory
compilation terminated.
Makefile:94: recipe for target sys.o failed
make[1]: *** [sys.o] Error 1
```

timholy committed:
Fix missing type parameter in _array_for

This fixes a performance regression in the lucompletepiv benchmark triggered by generalizing _array_for to work with indices. This call, first introduced in b363cc70ee6f09bb4472944f8690283793c6052a, likely always triggered dynamic method dispatch. The generalization to indices seemed to make that worse.

yuyichao committed:
A trick to fix continue/break/return handling in for loop testset.

Fix #17908

dhoegh committed:
Move env.jl test to separate test file

timholy committed:
Provide a more useful error message for calling map on a Dict with a non-Pair return type

dhoegh committed:
Fix #17956 and add test. The bug was introduced in 103db50aa6c67d8e732e8e2bd7f5743bae17e369

dhoegh committed:
Deprecate `delete!(ENV, k ,def)` as it is inherently type unstable and no other associative type implement the method.

dhoegh committed:
move doc for `withenv`

timholy committed:
Silence unneeded output in backtrace test

timholy committed:
Force inlining on indices(A, d)

vtjnash committed:
stop curl from rebuilding incessantly

vtjnash committed:
fix effect_free computation over getfield

fix #18015

Sacha0 committed:
Deprecate `@vectorize_1arg` and `@vectorize_2arg`.

Sacha0 committed:
Update documentation following deprecation of macro-vectorized functions and deprecation of `@vectorize_1arg` and `@vectorize_2arg` themselves. Also add cross-referenes between max and maximum, and also between min and minimum.

bicycle1885 committed:
fix a typo in Base.split (#18008)

* fix a typo in Base.split

* fixup! fix a typo in Base.split

tkelman committed:
Fix a row-indexing bug with sparse matrices

that have non-Int indices, introduced by #13612
and which has caused `make -C test/perf` to fail for the last 9 months

searchsortedfirst does not have methods for general Integer indices

KristofferC committed:
specialize on type in getindex for types

dhoegh committed:
Formatting: indent the OS specific blocks.

yuyichao committed:
Lock libmap and use atomic load/store on library handle/symbol GVs

Fix #18020

wookay committed:
Remove RNG from exports (#18023)

akc committed:
Fix a trivial typo in devdocs

tkoolen committed:
Make Colon immutable. Fixes 18034.

vtjnash committed:
test libccalltest.so with split debug info

timholy committed:
Make `find` indices-aware

timholy committed:
Fixes and more tests for indexing with non-1 logical arrays

vtjnash committed:
cleanup stacktraces method signatures

ensure that we are using consistent types
and introduce ProfileFormat to contain all of the user options

kshyatt committed:
Moved so many docs out of HelpDB. More examples.

vtjnash committed:
profile: add mincount argument

the min-count argument to Profile.print can be used to
filter out lines from the display that are insignificant

vtjnash committed:
fix "improved" correctness of fieldtype

the getfield_tfunc was missing a test for whether
the fields were all equivalent before concluding
that the result type was exact

fixes the fix #17953
fix #18037

JeffBezanson committed:
fall back to dynamic dispatch in union-splitting optimization

This should help #17932. If a function returns an unanticipated type
that wasn't part of the `Union` seen by inference, do dynamic
dispatch instead of erroring.

musm committed:
Update modules.rst adding type to show for v0.5 (#18018)

Update to add type annotation otherwise we get
```
ERROR: MethodError: no method matching display(::MyType)
```

mzaffalon committed:
added example of array type conversion (#17663)

* added example of array type conversion

... because it seems to bite less experienced users like myself.
(Stolen from Terry Seaward's email, https://groups.google.com/d/msg/julia-users/wzaJw_FfNlA/qcuZ4FggEL4J)

* added example of array type conversion

simonbyrne committed:
Document known problem with setrounding.

Fixes the doc problem mentioned in #17926, but not the underlying problem.

MatthiasJReisinger committed:
Simplify bounds checks for multi-dimensional array accesses

This simplifies the array bounds check code that is emitted for
multi-dimensional array accesses that use "regular" indexing, i.e.
accesses of the form `A[i1,i2,...,iN]` to some `N`-dimensional array
`A`.

For example, with this change, the access `A[i,j,k]` to an array
with the three dimensions `m`, `n` and `o` now leads to bounds checks
that correspond to the following pseudo code:

```
if (i >= m)
  out_of_bounds_error();
else if (j >= n)
  out_of_bounds_error();
else if (k >= o)
  out_of_bounds_error();
```

So far, the following more complicated bounds checks would have been
emitted:

```
if (i >= m)
  out_of_bounds_error();
else if (j >= n)
  out_of_bounds_error();
else if (((k * n + j) * m + i) >= m * n * o)
  out_of_bounds_error();
```

Julia also allows one-dimensional and "partial" linear indexing
(see #14770), i.e. the number of indices used to access an array does
not have to match the actual number of dimensions of the accessed array.
For this case we still have use this old scheme.

One motivation for this change was the following: expressions like
`((k * n + j) * m + i)` are non-affine and Polly would not be able
to analyze them. This change therefore also facilitates Polly's bounds
check elimination logic, which would hoist such checks out of loops
or may remove them entirely where possible.

maleadt committed:
Clear the IRBuilder's insertion point after emitting a function.

The function might get finalized, invalidating the IP.
However, in some cases this invalid IP may get saved and restored, accessing the invalid IP while doing so.

Example code path accessing an invalid IP:
 -> jl_cfunction_object (nested_compile=true, but doesn't change IP)
 -> gen_cfun_wrapper
 -> jl_compile_linfo (saves and restores invalid IP)

simonbyrne committed:
update docs, use new markdown syntax

simonbyrne committed:
wrap help text

omus committed:
Fix bug with operator precendence

pranavtbhat committed:
Define find and findnz for SparseVector (#18049)

JeffBezanson committed:
assume `getfield` is non-volatile if inferred to be constant

fixes regression caused by #18017

JeffBezanson committed:
fix #18051, allocation in sparse vector getindex

Caused by the return type of `ord()` being unknown.

maleadt committed:
ASAN: provide default options.

maleadt committed:
ASAN/MSAN: document use of sanitizers.

[ci skip]

maleadt committed:
Add test for #18054.

Note that this test requires a memory sanitizer (ASAN, valgrind) to detect failure.

musm committed:
Fix for consistency.

JeffBezanson committed:
fix #10633, remove redundant definitions for `map(Integer, a)` etc.

omus committed:
Improve URL regex

vtjnash committed:
improve FreeBSD / generic posix support

timholy committed:
Some indices generalizations for linalg/generic (#18032)

omus committed:
Correct invalid SSH URL

An SSH URL with a path uses a colon between the host and the path:

	git@github.com:JuliaLang/Example.jl (correct)
	git@github.com/JuliaLang/Example.jl (incorrect)

vtjnash committed:
move recheck_tuple_intersection to thread-local

vtjnash committed:
ensure in_typeinf_loop is only read by the thread with the typeinf lock

KristofferC committed:
do not catch everything in isassigned

tkelman committed:
Reinstate downloading old versions of winrpm gcc dll's (#18059)

This mostly reverts #17906 and puts #15521 back in place,
but from a slightly more permanent download location now.

Unfortunately using the latest opensuse copy of the gcc dll's
is causing issues when a cygwin-built Julia tries to load an
opensuse-built libzmq.dll. Using these slightly old gcc 5 versions
seems to fix it. I suspect the difference is which libstdc++ ABI
is being used. GCC 5 has both old and new available, but many distros
had it using the old ABI by default for compatibility. GCC 6 (which
opensuse is using now) is more likely to be using the new ABI by
default.

Keno committed:
In cgmemmgr, allow writing permissions on the read page for the debugger

The permissions specified in MapViewOfFile specify the maximum permissions
the file can ever have, which means that even permissions bypass by the
debugger is not allowed to write it. Instead, set the maximum permissions
to RWX and VirtualProtect down to the permissions we need. This is the same
behavior as we use on Mac/Linux.

omus committed:
Removed SSH test from libgit2-online

Test was only working for environments which had GitHub SSH keys setup.

omus committed:
Update URL parsing tests

Added a new explicit SSH protocol test which generated other changes.

vtjnash committed:
factor out the jl_get_llvmf code more cleanly

jl_get_llvmf was left in so that this can be backported to v0.5 without
altering the C api (in case it was used in a package)

also correct / add error handling and proper locks

kshyatt committed:
Add missing tests for diagonal and uniformscaling (#17921)

* Add missing tests for diagonal and uniformscaling

* Change A_mul_B to *

* Moved things out of nested loop that don't need to be there

kshyatt committed:
Add tests for flipdim, unary ops

kshyatt committed:
Add tests for iteration methods on numbers

vtjnash committed:
split static-eval to its individual users

most users of static-eval didn't need most of it's complexity
and didn't actually want its deprecation warning behavior

this also will make it easier to delete this function as the
users of it get fixed not to need it

yuyichao committed:
Improve arch/cpu detection/selection on ARM and AArch64

* Allow `cpu_target` to specify a generic arch, matching the behavior on x86
* Detect the CPU arch version with `uname`
* Require `armv6`

    Close #13270 (`armv5` is not supported)
    Fix #18042

* Remove warning about generic arch since it's not really useful

    Fix #17549

* Require at least the same ARM arch version and profile the C code is compiled with

mcprentiss committed:
updates to README.md (#18093)

This link has moved.

ViralBShah committed:
Build curl with minimal external libraries.
Link to our own libssh2.

kshyatt committed:
Added test for complex on SharedArrays (#18097)

yuyichao committed:
Fix compilation on llvm 3.9

staticfloat committed:
Set LIB_INSTALL_DIR for cmake-based projects (#18047)

* Set LIB_INSTALL_DIR for cmake-based projects

This fixes `MULTIARCH_INSTALL=1` installations that place libraries not
into `<prefix>/lib` but into `<prefix>/lib/<host-triplet>`, like Debian
and Ubuntu.  Without this change, `libgit2`, `libssh2` and `mbedtls`
would place libraries into `<prefix>/lib`, silently failing until the
Julia bootstrap process attempted to dlopen one of them.

* Apply `LIB_INSTALL_DIR` to `CMAKE_COMMON`

tkelman committed:
readme formatting

[ci skip]

tkelman committed:
fix "recommanded" typo

yuyichao committed:
Fix compiler warnings on 32bits

kshyatt committed:
Missing tests for strings (#18089)

JeffBezanson committed:
disable llvm optimizations in JIT when generating output

When saving native code we re-run optimization on the whole module
as part of object file emission, so it is not needed for JIT compiled
code. We are already passing -O0 for package precompilation; this
extends that benefit to the system image.

JeffBezanson committed:
fix #18085, segfault on method add in loop

Closures that are part of out-of-scope method adds were not lifted
out of top-level loops.

stevengj committed:
make dlopen more conservative about opening files the user didn't request (#18061)

* make dlopen more conservative about opening files the user didn't request

* update manual

* recognize \server\foo and /foo/bar absolute paths on windows, similar to isabspath

* simplify isabspath from init.c and use it in dlopen instead of duplicating

dbeach24 committed:
Bugfix #18073 (#18105)

fix #18073: `any([])` and `all([])` (#18105)
(cherry picked from commit d259be55130f9d695262c3e528697c451cbb4e9e)

JeffBezanson committed:
fix more of the #18017 regression

andreasnoack committed:
Add istriu, istril, issymmetric, and ishermitian methods for UniformScaling (#17950)

JeffBezanson committed:
fix #18129, optimize some more cases of captured variables

kshyatt committed:
Missing tests for intfuncs

yuyichao committed:
Use unique identifier to avoid incorrect ditype reuse

kshyatt committed:
Tests for missing abstract array utils

pwl committed:
rational powers (fixes #18114) (#18118)

Fixes precision problem for rational powers of BigFloats (#18114)

tkelman committed:
Install all 3 mbedtls dlls on windows

and build tests

tkelman committed:
disable libpsl in curl build

should fix clean build of master

indent line continuations

kshyatt committed:
More tests for osnames (#18121)

kshyatt committed:
Two missing tests for intset (#18128)

timholy committed:
Make similar faster and safer (#18107)

* Make similar faster and safer

By converting the eltype argument into a ::Type{T} parameter, we avoid runtime method lookup.

Perhaps more importantly, the introduction of NeedsShaping makes `similar` safer, by preventing any possibility of an infinite recursion.

* Test that similar throws a MethodError for unsupported dims types

MichaelHatherly committed:
Fix incorrect code block in rounding modes docs (#18137)

The `::` syntax appearing before the `doctest::` block was causing
the doctest not be rendered correctly.

tkelman committed:
disable using zlib in mbedtls, libssh2, and libcurl

since it requires additional dev headers
fix #18101

andreasnoack committed:
Remove +-(UniformScaling, Number) to avoid ambiguous behavior as pointed (#17607)

out in #17083

maleadt committed:
Expose MEMDEBUG, and use it to fix the #14173 test

vtjnash committed:
fix sort / fallback order for incremental precompile

fix #18069

JeffBezanson committed:
replace branch in bit shift operators, helps #18135

stevengj committed:
make sure .= uses Base.identity, not a local identity (#18122)

vtjnash committed:
change incremental serialize to use deferred errors

this gives the caller more control over when an how errors get printed

vtjnash committed:
keep track of full dependency state while building incremental precompile caches

this ensures that we only recompile a dependency if either:
it can be loaded into the current session, or
the user explicitly requests it (via reload)

thereby reducing the occurrences and improving the accuracy
of the "uuid didn't match" error message :)

tkelman committed:
Cleanup: replace misleading STAGE#_DEPS variables with DEP_LIBS

tkelman committed:
Improve handling of curl, mbedtls, libssh2 USE_SYSTEM

fix #17884

tkelman committed:
Set CMAKE_INSTALL_LIBDIR and CMAKE_INSTALL_BINDIR

in CMAKE_COMMON, so libssh2 installs without an x86_64-linux-gnu
multiarch prefix

tkelman committed:
Disable building static mbedtls library

since we don't need it

Go back to just copying the mbedtls dll even in cygwin-cross,
otherwise it puts .dll.a files that we don't need alongside the dlls
(should probably fix this upstream)

tkelman committed:
Also just copy libgit2 dll for its install target

since otherwise it puts pkg-config files in LIB_INSTALL_DIR
which we don't need - if mbedtls separates runtime from library
destinations then we could separate BIN_INSTALL_DIR from LIB_INSTALL_DIR
or convince libgit2 to use the standard GNUInstallDirs names

yuyichao committed:
Do codegen in control flow order

* Fix codegen of `SSAValue` whose assignment appears later than the user
  in syntactic order due to backward branches.
* Disallow more expression types in value position
* Use improved debug info metadata in coverage and allocation tracking too

tkelman committed:
Fix indent in sparsevector.jl

and remove some unnecessary blank lines
[ci skip]

yuyichao committed:
Try fixing LLVM 3.3 compilation (#18138)

ranjanan committed:
Add test for 15913 and 15896 (#17991)

* Add test for 15913 and 15896

* Change name of test

twadleigh committed:
Fixes #18141, eliminating null in value when iterating over `ENV`. (#18144)

* Eliminate trailing null from iterator over `ENV`.

Fixes #18141.

* Add failing test for #18141.

timholy committed:
reshape(::Array, Val{N}) always returns an Array

dbeach24 committed:
fast rand(::Dict) and rand(::Set) (#18155)

fixes #16231

timholy committed:
reshape: only call to_shape when it will return Dims

MichaelHatherly committed:
Add note about docstrings for aliases (#18157)

As mentioned in https://github.com/JuliaLang/julia/pull/18041#issuecomment-241189817
docstrings attached to aliases of functions should be avoided where possible.

Keno committed:
Misc build fixes for win32 clang (#18164)

maleadt committed:
Increase MEMDEBUG allocation threshold for #14173 test

timholy committed:
Add do-block support for redirect_std[out,err,in]

Fixes #7022. In particular, these simplify testing for warnings.

yuyichao committed:
Make sure `:push_loc` meta always has a corresponding `:pop_loc`

* Preserve meta node and line number info during dead code elimination
* Insert `:push_loc` and `:pop_loc` in pairs during lowering

Fix #16578

maleadt committed:
ASAN: disable leak detection for flisp.

timholy committed:
Fix fft(::Vector{Complex{Int}})

JeffBezanson committed:
fix #18173, closure lowering error caused by #18130

vtjnash committed:
document __init__ behavior during precompile

fix #18115

vtjnash committed:
incremental deserialize: handle LambdaInfo identity uniquing

this works to avoid having `Expr(:invoke)` creating
unintentional copies of LambdaInfo objects when they show
up in the system image

fix #18184

vtjnash committed:
incremental deserialize: optimize many simple common cases of flagref_list

many types are easy to compute during deserialization,
and don't need the full complexity of the flagref list to handle them
some of these are easy to detect ahead-of-time,
saving a small amount of deserialization effort

vtjnash committed:
precompile: major LLVM speed improvement

fix #15048

Commit:
typemap: handle vararg tuple subtyping in jl_typemap_assoc_by_type

Fixes a crash while loading ASTInterpreter

tkelman committed:
Fix the part of #18164 that broke the build

of cmake projects with gcc - CC_ARG and CXX_ARG are getting double processed
somewhere, so cmake is trying to call the compiler as '/usr/bin/gcc gcc'

add cut -s

kshyatt committed:
Moved docstrings to deal with aliasing

Commit:
ccalltest.c: ifndef Intel compiler for complex int (#18193)

Complex int is GNU extension that Intel compiler does not
support currently. #17145

eschnett committed:
Describe negative bit shift counts in NEWS.md (#18188)

vtjnash committed:
profile: add noisefloor argument

the noise-floor argument to Profile.print can be used to
filter out lines from the display that are insignificant,
based on the heuristic `n <= noisefloor * sqrt(N)`

stevengj committed:
inline native Julia scalars encountered in f.(args...) AST

yuyichao committed:
Pop noinline meta before label reindexing

martinholters committed:
Move specialized isequal definitions to avoid hitting #265

Fixes #18213.

yuyichao committed:
Store `propragate_inbounds` in lambda info

Move unexported `jl_sym_t*` GVs from `julia.h` to `julia_internal.h`

yuyichao committed:
Add void_use_elim_pass!

vtjnash committed:
inliner needs to do method match tests early

fix #18222

stevengj committed:
improve string docs to define String etc (#18216)

* improve string docs to define String etc

* tweak

MichaelHatherly committed:
Don't require blank line before markdown lists

Fixes #11249 to follow the CommonMark spec for lists which does not
require a blank line prior to the start of a list.

stevengj committed:
fix typo in ##18205

andreasnoack committed:
Fix chol for Symmetric(A, :L). (#18142)

andreasnoack committed:
Don't throw in eigs on semidefinite Bs for generalized eigenproblems. (#17873)

This makes it possible to solve problems with semidefinite B via explicit
shift.

rened committed:
add missing "Threads." in example (#18219)

yuyichao committed:
Handle `VecElement` in `julia_const_to_llvm`

Also clean up `julia_const_to_llvm` to avoid boxing due to `getfield`.

Fix #18236

Sacha0 committed:
Deprecate methods that convert from Diagonal and Bidiagonal to <:AbstractTriangular. Remove tests of those convert methods. (#17723)

tkelman committed:
hopefully unbreak osx travis llvm linking (#18241)

ref https://github.com/staticfloat/homebrew-julia/commit/f57dcc20464c9ce4758a6fca088164a591daf1cd#diff-633fd9f8448015391714f35e04d39b5bL154
[av skip]

eschnett committed:
Clarifiy performance results (#18242)

Closes #18231.

yuyichao committed:
Improve inlining cost model

Consistently ignore line numbers and metadata nodes in all cases.

yuyichao committed:
Clean up meta expression detection/handling

yuyichao committed:
Handle boundscheck elimination in type inference

Also collapse redundant debug info when coverage is off.
Fix incorrect `Expr(:inbounds, false)` elimination.

yuyichao committed:
Delete boundscheck handling from codegen

timholy committed:
Speed up indices for SubArray

timholy committed:
Use inbounds in reductions with generic iterators

KristofferC committed:
add faster done for strings (#18259)

martinholters committed:
Test for Dict not triggering a deprecation due to ::Integer==::Char comparison (#18234)

See #18213.

garrettthomaskth committed:
lufact for sparse matrix pivot option error (#18246)

* lufact for sparse matrix pivot option error

base/sparse/umfpack.jl includes the following method definition for lufact

`lufact(A::SparseMatrixCSC, pivot::Type{Val{false}}) = lufact(A)`

This should likely be

`lufact(A::SparseMatrixCSC, pivot::Type{Val{true}}) = lufact(A)`

because in lufact pivoting is on by default.

The error is shown in the following example

```
A = speye(4)
A[1:2,1:2] = [-.01 -200; 200 .001]
F = lufact(A,Val{false})
F[:p]
```
which returns
```
julia> F[:q]
4-element Array{Int64,1}:
 3
 4
 1
 2
```
However it should return
```
julia> F[:q]
4-element Array{Int64,1}:
1
2
3
4
```
because pivoting was turned off.

* Added test for #18246 and #18244

* 4 space indent

oops my bad

* remove unnecessary lufact method

* update test for removed lufact method definition

KristofferC committed:
Revert #182599 "add faster done for strings" (#18275)

* Revert "add faster done for strings (#18259)"

This reverts commit 30ee10b0fff2cf7c0f14ba5531b8f980a7f5c975.

* add test that failed with  #18259

Evizero committed:
Add missing convert(Vector, ...) and convert(Matrix, ...) methods (#17848)

This commit will allow a user to call:
convert(Vector, my_abstractvector_of_eltype_Foo)
convert(Matrix, my_abstractmatrix_of_eltype_Foo)

which is useful because
- for the case that the variable is a subarray of the same dimensionality
- for consistency with PR #17066

p-zubieta committed:
Some broadcast fixes (#18200)

* Fix #18176 (broadcast over mixtures of arrays and numeric scalars)

* Fix #17984 (broadcast behavior over empty arrays)

tkelman committed:
Rename the FooBar type in test/core.jl (#18279)

so `JULIA_CPU_CORES=1 ./julia test/runtests.jl core compile` passes
after #18150

timholy committed:
Add a more efficient implementation of in(::CartesianIndex, ::CartesianRange) (#18277)

nalimilan committed:
Add tests for conversions from ReshapedArray to Array/Vector/Matrix (#18262)

These were not tested systematically, and are more generally useful to
test conversions from AbstractArray.

twadleigh committed:
Install libopenlibm.dll.a in windows. Bump openlibm to v0.5.4. (#18265)

staticfloat committed:
Small build system modifications to make ppc64le compileable (#18258)

* Small build system modifications to make ppc64le compileable

* Make things more explicit for big-endian PPC64

helgee committed:
Fix deprecated flag in Pkg tests. (#18283)

kshyatt committed:
Add tests for Hermitian cholesky methods (#18284)

jdlangs committed:
Remove comments from same line of variable declarations in Make.inc

ViralBShah committed:
ppc64 build cleanups

stevengj committed:
tiny cleanup (#18233)

just a random thing I noticed

stevengj committed:
display something useful for text/plain output of invalid String (#18296)

andreasnoack commented:
@mfasi It would be great to get this into Julia. Would you have time to finish the PR?

Sacha0 committed:
Revise a few vectorized-math methods, specifically eliminating temporaries and enabling broadcast loop fusion.

Sacha0 committed:
Remove now-redundant integer-specialized condskeel methods.

kshyatt committed:
More wording improvements, few more doctests

tkelman committed:
set JULIA_CPU_TARGET default after MARCH-dependent block

otherwise #18258 causes "Target architecture mismatch" errors in generic linux binaries

ViralBShah committed:
Update OpenBLAS to v0.2.19 for Power8.

stevengj committed:
rm stray : in manual

andreasnoack committed:
Avoid aliasing in in UniformScaling*AbstractMatrix (#18286)

* Avoid aliasing in in UniformScaling*AbstractMatrix

...and remove unnecessary UniformScaling*SparseMatrixCSC methods

* Broaden the tests for non-commutative multiplication

* Add Quaternion test case for q*[q] and clean up the Quaternion test type

twadleigh committed:
Repeatedly divide read buffer size by 8 until success. Fixes #11481.

Dividing by 8 leads to success on the second try in the case of the failure of --lisp on Windows 7.

BrandonEdwards committed:
Fix #18189 (Add deprecation of  @unix_only, etc. to NEWS) (#18254)

* Fix #18189 (Add deprecation of  etc. to NEWS)

* Remove mentioning of @static, add as a note

* Added reference to Handling Operating System Variation section in manual

* Changed reference numbers

* Change PR number

yuyichao committed:
Fix `jl_get_specialization` signature mismatch

Also fix a few other compiler warnings.

vtjnash committed:
make compile fast

we really badly need a heuristic for detecting simple methods like this.

matthieugomez committed:
Update utf8proc.jl (#18327)

add ! suffix to isgraphemebreak! to clarify that the state is changed

simonster committed:
Fix docs for isdefined

stevengj committed:
don't throw DomainError for negative integer powers of ±1 (#18342)

* don't throw DomainError for negative integer powers of ±1

* fix replutil test that was expecting 1^(-1) to fail

stevengj commented:
Uh oh, looks like something went terribly wrong with a rebase...

mfasi commented:
@stevengj Yes, you're right. No Idea how to fix this (I'm trying but it looks like I went a git push -f too far), I think I will have to start over with a new branch

alyst commented:
@mfasi To avoid making a new PR you can start a new local branch from the current master, cherry-pick the relevant commits from this branch, make sure that it works for you locally, then replace this github branch with your new local one.

andreasnoack commented:
@mfasi Did you ever manage to recover this work on a new branch? As mentioned, we'd be very interested in getting this into Julia.

stevengj commented:
Merged in #21184.
